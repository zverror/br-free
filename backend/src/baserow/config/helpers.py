import asyncio

from loguru import logger


class dummy_context:
    async def __aenter__(self):
        pass

    async def __aexit__(self, exc_type, exc, traceback):
        pass


class ConcurrencyLimiterASGI:
    """
    Helper wrapper on ASGI app to limit the number of requests handled
    at the same time.
    """

    def __init__(self, app, max_concurrency: int | None = None):
        self.app = app
        logger.info(f"Setting ASGI app concurrency to {max_concurrency}")
        self.semaphore = (
            asyncio.Semaphore(max_concurrency)
            if (isinstance(max_concurrency, int) and max_concurrency > 0)
            else dummy_context()
        )

    async def __call__(self, scope, receive, send):
        async with self.semaphore:
            await self.app(scope, receive, send)
