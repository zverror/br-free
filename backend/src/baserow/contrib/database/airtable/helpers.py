from typing import List

from baserow.contrib.database.fields.models import Field, SelectOption

from .constants import AIRTABLE_BASEROW_COLOR_MAPPING


def import_airtable_date_type_options(type_options) -> dict:
    date_format_mapping = {"European": "EU", "US": "US"}
    time_format_mapping = {"12hour": "12", "24hour": "24"}
    return {
        "date_format": date_format_mapping.get(type_options.get("dateFormat"), "ISO"),
        "date_include_time": type_options.get("isDateTime", False),
        "date_time_format": time_format_mapping.get(
            type_options.get("timeFormat"), "24"
        ),
    }


def import_airtable_choices(field_id: str, type_options: dict) -> List[SelectOption]:
    order = type_options.get("choiceOrder", [])
    choices = type_options.get("choices", [])
    return [
        SelectOption(
            # Combine select id with choice id as choice id is not guaranteed to be
            # unique across table
            id=f"{field_id}_{choice['id']}",
            value=choice["name"],
            color=AIRTABLE_BASEROW_COLOR_MAPPING.get(
                # The color isn't always provided, hence the fallback to an empty
                # color, which will fallback on the blue color.
                choice.get("color", ""),
                "blue",
            ),
            order=order.index(choice["id"]),
        )
        for choice in choices.values()
    ]


def set_select_options_on_field(
    field: Field, field_id: str, type_options: dict
) -> Field:
    """
    Set the `select_options` of a field in the prefetched objects cache. Overriding
    this object makes sure that when later `field.select_options.all()` is executed,
    it will return Airtable choices. This for example happens in
    `FieldType::export_serialized`.

    :param field: The field where the select must be set on.
    :param field_id: id of airtable choice field
    :param type_options: The options where to extract the Airtable choices from.
    :return: The updated field object.
    """

    # Overriding this object makes sure that when later
    # `field.select_options.all()` is executed, it will return Airtable choices.
    # This for example happens in `FieldType::export_serialized`.
    field._prefetched_objects_cache = {
        "select_options": import_airtable_choices(field_id, type_options)
    }
    return field
