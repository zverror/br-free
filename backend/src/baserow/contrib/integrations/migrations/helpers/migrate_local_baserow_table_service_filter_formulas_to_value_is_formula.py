"""
Helpers for 0006_migrate_local_baserow_table_service_filter_formulas_to_value_is_formula
"""

from baserow.core.formula import BaserowFormulaSyntaxError, get_parse_tree_for_formula


def value_parses_as_formula(value: str) -> bool:
    """
    Returns whether the service filter value is a formula or not.
    """

    # If the value is empty, it's not a formula.
    if not value:
        return False

    # If it's a string that's an integer, it's not a formula.
    try:
        int(value)
        return False
    except ValueError:
        pass

    try:
        get_parse_tree_for_formula(value)
        return True
    except BaserowFormulaSyntaxError:
        return False
