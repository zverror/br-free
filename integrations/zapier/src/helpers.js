const { unsupportedBaserowFieldTypes } = require('./constants')

/**
 * Fetches the fields of a table and converts them to an array with valid Zapier
 * field objects.
 */
const getRowInputValues = async (z, bundle) => {
  if (!bundle.inputData.tableID) {
    throw new Error('The `tableID` must be provided.')
  }

  const fieldsGetRequest = await z.request({
    url: `${bundle.authData.apiURL}/api/database/fields/table/${bundle.inputData.tableID}/`,
    method: 'GET',
    headers: {
      'Accept': 'application/json',
      'Authorization': `Token ${bundle.authData.apiToken}`,
    },
  })

  return fieldsGetRequest.json.map(v => {
    return mapBaserowFieldTypesToZapierTypes(v)
  })
}

/**
 * Fetches the fields and converts the input data to Baserow row compatible data.
 */
const prepareInputDataForBaserow = async (z, bundle) => {
  if (!bundle.inputData.tableID) {
    throw new Error('The `tableID` must be provided.')
  }

  const fieldsGetRequest = await z.request({
    url: `${bundle.authData.apiURL}/api/database/fields/table/${bundle.inputData.tableID}/`,
    method: 'GET',
    headers: {
      'Accept': 'application/json',
      'Authorization': `Token ${bundle.authData.apiToken}`,
    },
  })

  let rowData = { id: bundle.inputData.rowID }
  fieldsGetRequest
    .json
    .filter(
      (baserowField) =>
        baserowField.read_only
          || !unsupportedBaserowFieldTypes.includes(baserowField.type)
    )
    .filter((baserowField) => bundle.inputData.hasOwnProperty(baserowField.name))
    .forEach(baserowField => {
      let value = bundle.inputData[baserowField.name]

      if (baserowField.type === 'multiple_collaborators') {
        value = value.map(id => {
          return { id }}
        )
      }

      rowData[baserowField.name] = value
    })

  return rowData
}

/**
 * Converts the provided Baserow field type object to a Zapier compatible object.
 */
const mapBaserowFieldTypesToZapierTypes = (baserowField) => {
  const zapType = {
    key: baserowField.name,
    label: baserowField.name,
    type: 'string'
  }

  if (baserowField.type === 'long_text') {
    zapType.type = 'text'
  }

  if (baserowField.type === 'boolean') {
    zapType.type = 'boolean'
  }

  if (baserowField.type === 'number') {
    zapType.type = 'integer'

    if (baserowField.number_decimal_places > 0) {
      zapType.type = 'float'
    }
  }

  if (baserowField.type === 'boolean') {
    zapType.type = 'boolean'
  }

  if (baserowField.type === 'rating') {
    zapType.type = 'integer'
  }

  if (['single_select', 'multiple_select'].includes(baserowField.type)) {
    const choices = {}
    baserowField.select_options.forEach(el => {
      choices[`${el.id}`] = el.value
    })
    zapType.type = 'string'
    zapType.choices = choices
  }

  if (baserowField.type === 'multiple_select') {
    zapType.list = true
  }

  if (baserowField.type === 'link_row') {
    zapType.type = 'integer'
    zapType.helpText = 'Provide row ids that you want to link to.'
    zapType.list = true
  }

  if (baserowField.type === 'multiple_collaborators') {
    zapType.type = 'integer'
    zapType.helpText = 'Provide user ids that you want to link to.'
    zapType.list = true
  }

  if (baserowField.type === 'date' && !baserowField.date_include_time) {
    zapType.type = 'date'
    zapType.helpText =
      'the date fields accepts a date in ISO format (e.g. 2020-01-01)'
  }

  if (baserowField.type === 'date' && baserowField.date_include_time) {
    zapType.type = 'datetime'
    zapType.helpText =
      'the date fields accepts date and time in ISO format (e.g. 2020-01-01 12:00)'
  }

  if (
    baserowField.read_only
    || unsupportedBaserowFieldTypes.includes(baserowField.type)
  ) {
    // Read only and the file field are not supported.
    return
  }

  return zapType
}

module.exports = {
  getRowInputValues,
  prepareInputDataForBaserow,
  mapBaserowFieldTypesToZapierTypes,
}
