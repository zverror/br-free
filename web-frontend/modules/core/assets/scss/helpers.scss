%clearfix {
  &::after {
    clear: both;
    content: '';
    display: table;
  }
}

.clearfix {
  @extend %clearfix;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.background-white {
  background-color: $white;
}

.color-primary {
  color: $color-primary-500 !important;
}

.color-success {
  color: $color-success-500 !important;
}

.color-warning {
  color: $color-warning-500 !important;
}

.color-error {
  color: $color-error-500 !important;
}

.color-neutral {
  color: $color-neutral-500 !important;
}

.padding-top-2 {
  padding-top: 16px !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-24 {
  margin-top: 24px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottom-4 {
  margin-bottom: 40px !important;
}

.margin-left-auto {
  margin-left: auto !important;
}

.margin-left-0 {
  margin-left: 0 !important;
}

.margin-left-1 {
  margin-left: 8px !important;
}

.margin-left-2 {
  margin-left: 16px !important;
}

.margin-left-3 {
  margin-left: 30px !important;
}

.margin-left-4 {
  margin-left: 40px !important;
}

.margin-right-auto {
  margin-right: auto !important;
}

.margin-right-0 {
  margin-right: 0 !important;
}

.margin-right-1 {
  margin-right: 8px !important;
}

.margin-right-2 {
  margin-right: 16px !important;
}

.margin-right-3 {
  margin-right: 30px !important;
}

.margin-right-4 {
  margin-right: 40px !important;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.forced-text-overflow-initial {
  text-overflow: initial !important;
}

.forced-pointer-events-auto {
  pointer-events: auto !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

.whitespace-pre-wrap {
  white-space: pre-wrap;
}

.remove-number-input-controls,
%remove-number-input-controls {
  appearance: textfield !important;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    appearance: none;
    margin: 0;
  }
}

.flex {
  width: 100%;
  display: flex;
  gap: var(--gap, 10px);
}

.flex-wrap {
  flex-wrap: wrap;
}

.flex-100 {
  width: 100%;
  flex-basis: 100%;
}

.flex-grow-1 {
  flex-grow: 1;
}

.align-items-center {
  align-items: center;
}

.justify-content-space-between {
  justify-content: space-between;
}

.position-relative {
  position: relative;
}

.inline-block {
  display: inline-block;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}

/* spacer classes generation */

@each $size in $spacings {
  .p-#{$size},
  %p-#{$size} {
    padding: #{$size}px;
  }

  .px-#{$size},
  %px-#{$size} {
    padding-left: #{$size}px;
    padding-right: #{$size}px;
  }

  .py-#{$size},
  %py-#{$size} {
    padding-top: #{$size}px;
    padding-bottom: #{$size}px;
  }

  .m-#{$size},
  %m-#{$size} {
    margin: #{$size}px;
  }

  .my-#{$size},
  %my-#{$size} {
    margin-top: #{$size}px;
    margin-bottom: #{$size}px;
  }

  .mx-#{$size},
  %mx-#{$size} {
    margin-left: #{$size}px;
    margin-right: #{$size}px;
  }

  .pl-#{$size},
  %pl-#{$size} {
    padding-left: #{$size}px;
  }

  .ml-#{$size},
  %ml-#{$size} {
    margin-left: #{$size}px;
  }

  .pr-#{$size},
  %pr-#{$size} {
    padding-right: #{$size}px;
  }

  .mr-#{$size},
  %mr-#{$size} {
    margin-right: #{$size}px;
  }

  .pb-#{$size},
  %pb-#{$size} {
    padding-bottom: #{$size}px;
  }

  .mb-#{$size},
  %mb-#{$size} {
    margin-bottom: #{$size}px;
  }

  .pt-#{$size},
  %pt-#{$size} {
    padding-top: #{$size}px;
  }

  .mt-#{$size},
  %mt-#{$size} {
    margin-top: #{$size}px;
  }
}
