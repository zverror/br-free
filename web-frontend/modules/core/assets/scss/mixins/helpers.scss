@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin flex-align-items($gap: 0, $display: flex) {
  display: $display;
  align-items: center;
  gap: $gap;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if $top != null and $right == null and $bottom == null and $left == null {
    inset: $top $top $top $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if $top != null and $right != null and $bottom == null and $left == null
  {
    inset: $top $right $top $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top {
      top: $top;
    }

    @if $right {
      right: $right;
    }

    @if $bottom {
      bottom: $bottom;
    }

    @if $left {
      left: $left;
    }
  }
}
