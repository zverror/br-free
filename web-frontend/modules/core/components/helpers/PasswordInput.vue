<!--
This helper component can only be used in a form which uses the password validation
from modules/core/validators.js
as ist expects the following vuelidate validation state params

- required
- maxLength
- minLength

In order to use this component in another component/form it needs a value
prop which can be passed via the 'v-model' directive. That means the state for the 
password value always lives in the parent component.
The parent component also needs to provide the vuelidate password validation in order
for this component to show all the possible validation errors.

Example Usage:

<PasswordInput
   v-model="PASSWORDSTATE"
   :validation-state="the vuelidate validation instance of the given password value"
/>

In case the password validation rules change, the passwordValidation object in the
validators file needs to be updated, as well as this component.

modules/core/validators.js

-->
<template>
  <FormGroup
    small-label
    :label="label"
    required
    :error="validationState.$error"
  >
    <FormInput
      :error="validationState.$error"
      type="password"
      size="large"
      :autocomplete="autocomplete"
      :value="value"
      :placeholder="placeholder"
      @blur="validationState.$touch()"
      @input="$emit('input', $event)"
    >
    </FormInput>

    <template #error>
      <span v-if="validationState.$error && !validationState.required">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{ $t('error.inputRequired') }}
      </span>
      <span v-if="validationState.$error && !validationState.maxLength">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{
          $t('error.maxLength', {
            max: validationState.$params.maxLength.max,
          })
        }}
      </span>
      <span v-if="validationState.$error && !validationState.minLength">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{
          $t('error.minLength', {
            min: validationState.$params.minLength.min,
          })
        }}
      </span>
    </template>
  </FormGroup>
</template>

<script>
export default {
  name: 'PasswordInput',
  props: {
    validationState: {
      type: Object,
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: false,
      default: null,
    },
    autocomplete: {
      type: String,
      required: false,
      default: 'new-password',
    },
    placeholder: {
      type: String,
      required: false,
      default: null,
    },
    errorPlaceholderClass: {
      type: String,
      required: false,
      default: '',
    },
    showErrorIcon: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
}
</script>
