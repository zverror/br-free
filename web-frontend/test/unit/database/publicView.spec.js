import { TestApp } from '@baserow/test/helpers/testApp'
import PublicGrid from '@baserow/modules/database/pages/publicView'

// Mock out debounce so we dont have to wait or simulate waiting for the various
// debounces in the search functionality.
jest.mock('lodash/debounce', () => jest.fn((fn) => fn))

describe('Public View Page Tests', () => {
  let testApp = null
  let mockServer = null

  beforeAll(() => {
    testApp = new TestApp()
    mockServer = testApp.mockServer
  })

  afterEach(() => testApp.afterEach())

  test('Can see a publicly shared grid view', async () => {
    const slug = 'testSlug'
    const gridViewName = 'my public grid view name'
    givenAPubliclySharedGridViewWithSlug(gridViewName, slug)

    const publicGridViewPage = await testApp.mount(PublicGrid, {
      asyncDataParams: {
        slug,
      },
    })

    expect(publicGridViewPage.html()).toContain(gridViewName)
    expect(publicGridViewPage.element).toMatchSnapshot()
  })

  test('Publicly shared view is not saved as a last visited view', async () => {
    const slug = 'testSlug'
    const gridViewName = 'my public grid view name'
    givenAPubliclySharedGridViewWithSlug(gridViewName, slug)

    await testApp.mount(PublicGrid, {
      asyncDataParams: {
        slug,
      },
    })

    const allCookies = testApp.store.$cookies
    const cookieValue = allCookies.get('defaultViewId')
    expect(cookieValue.length).toBe(0)
  })

  function givenAPubliclySharedGridViewWithSlug(name, slug) {
    const fields = [
      {
        id: 0,
        name: 'Name',
        type: 'text',
        primary: true,
      },
      {
        id: 1,
        name: 'Last name',
        type: 'text',
      },
      {
        id: 2,
        name: 'Notes',
        type: 'long_text',
      },
      {
        id: 3,
        name: 'Active',
        type: 'boolean',
      },
    ]
    const gridView = mockServer.createPublicGridView(slug, {
      name,
      fields,
    })
    mockServer.createPublicGridViewRows(slug, fields, [
      {
        id: 1,
        order: 0,
        field_1: 'name',
        field_2: 'last_name',
        field_3: 'notes',
        field_4: false,
      },
    ])

    return { gridView }
  }
})
