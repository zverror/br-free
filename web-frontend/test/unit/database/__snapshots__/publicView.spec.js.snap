// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Public View Page Tests Can see a publicly shared grid view 1`] = `
<div>
  <div>
    <div
      class="toasts__container-top"
    >
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
    </div>
     
    <div
      class="toasts__container-bottom"
    >
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
    </div>
  </div>
   
  <div
    class="public-view__table"
  >
    <div>
      <header
        class="layout__col-2-1 header"
      >
        <div
          class="header__loading"
          style="display: none;"
        />
         
        <ul
          class="header__filter"
        >
          <li
            class="header__filter-item"
          >
            <a
              class="header__filter-logo"
              href="https://baserow.io"
              target="_blank"
              title="Baserow - open source no-code database tool and Airtable alternative"
            >
              <div
                class="logo"
              >
                <img
                  alt="Baserow - open source no-code database tool and Airtable alternative"
                  class=""
                  height="14"
                  src="@baserow/modules/core/static/img/logo.svg"
                />
              </div>
            </a>
          </li>
           
          <li
            class="header__filter-item header__filter-item--grids"
          >
            <a
              class="header__filter-link header__filter-link--disabled"
            >
              <i
                class="header__filter-icon header-filter-icon--view color-primary iconoir-menu"
              />
               
              <span
                class="header__filter-name header__filter-name--forced"
              >
                <span
                  class=""
                  contenteditable="false"
                  placeholder=""
                >
                  my public grid view name
                </span>
              </span>
               
              <i
                class="header__sub-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <!---->
          </li>
           
          <!---->
           
          <li
            class="header__filter-item"
          >
            <div>
              <a
                class="header__filter-link"
              >
                <i
                  class="header__filter-icon iconoir-filter"
                />
                 
                <span
                  class="header__filter-name"
                >
                  viewFilter.filter - 0
                </span>
              </a>
               
            </div>
          </li>
           
          <li
            class="header__filter-item"
          >
            <div>
              <a
                class="header__filter-link"
              >
                <i
                  class="header__filter-icon iconoir-sort"
                />
                 
                <span
                  class="header__filter-name"
                >
                  viewSort.sort - 0
                </span>
              </a>
               
            </div>
          </li>
           
          <li
            class="header__filter-item"
          >
            <div>
              <a
                class="header__filter-link"
              >
                <i
                  class="header__filter-icon iconoir-book-stack"
                />
                 
                <span
                  class="header__filter-name"
                >
                  viewGroupBy.groupBy - 0
                </span>
              </a>
               
            </div>
          </li>
           
          <!---->
           
          <!---->
        </ul>
         
        <ul
          class="header__filter header__filter--full-width"
          table="[object Object]"
        >
          <li
            class="header__filter-item"
          >
            <div>
              <a
                class="header__filter-link"
              >
                <i
                  class="header__filter-icon iconoir-eye-off"
                />
                 
                <span
                  class="header__filter-name"
                >
                  gridViewHide.hideField - 0
                </span>
              </a>
               
            </div>
          </li>
           
          <li
            class="header__filter-item"
          >
            <div>
              <a
                class="header__filter-link"
              >
                <i
                  class="header__filter-icon iconoir-table-rows"
                />
                 
                <span
                  class="header__filter-name"
                >
                  gridViewRowHeight.name
                </span>
              </a>
               
              <div
                class="dropdown dropdown--floating"
                tabindex="0"
              >
                <!---->
                 
                <div
                  class="dropdown__items hidden"
                >
                  <!---->
                   
                  <ul
                    class="select__items"
                    tabindex="-1"
                  >
                    <li
                      class="select__item select__item--no-options"
                    >
                      <a
                        class="select__item-link"
                      >
                        <div
                          class="select__item-name"
                        >
                          <!---->
                           
                          <!---->
                           
                          <!---->
                           
                          <span
                            class="select__item-name-text"
                            title="gridViewRowHeight.small"
                          >
                            gridViewRowHeight.small
                          </span>
                        </div>
                         
                        <!---->
                      </a>
                       
                      <i
                        class="select__item-active-icon iconoir-check"
                      />
                    </li>
                     
                    <li
                      class="select__item select__item--no-options"
                    >
                      <a
                        class="select__item-link"
                      >
                        <div
                          class="select__item-name"
                        >
                          <!---->
                           
                          <!---->
                           
                          <!---->
                           
                          <span
                            class="select__item-name-text"
                            title="gridViewRowHeight.medium"
                          >
                            gridViewRowHeight.medium
                          </span>
                        </div>
                         
                        <!---->
                      </a>
                       
                      <i
                        class="select__item-active-icon iconoir-check"
                      />
                    </li>
                     
                    <li
                      class="select__item select__item--no-options"
                    >
                      <a
                        class="select__item-link"
                      >
                        <div
                          class="select__item-name"
                        >
                          <!---->
                           
                          <!---->
                           
                          <!---->
                           
                          <span
                            class="select__item-name-text"
                            title="gridViewRowHeight.large"
                          >
                            gridViewRowHeight.large
                          </span>
                        </div>
                         
                        <!---->
                      </a>
                       
                      <i
                        class="select__item-active-icon iconoir-check"
                      />
                    </li>
                  </ul>
                   
                  <!---->
                   
                  <!---->
                </div>
              </div>
            </div>
          </li>
           
          <li
            class="header__filter-item header__filter-item--right"
          >
            <div>
              <a
                class="header__filter-link"
              >
                <i
                  class="header__search-icon iconoir-search"
                />
                
    
  
              </a>
               
            </div>
          </li>
        </ul>
      </header>
       
      <div
        class="layout__col-2-2 content"
      >
        <div
          class="grid-view grid-view--row-height-undefined"
        >
          <div
            class="scrollbars"
            style="left: 72px;"
          >
            <!---->
             
            <!---->
          </div>
           
          <div
            class="grid-view__left"
            style="width: 72px;"
          >
              
            <div
              class="grid-view__inner"
              style="min-width: 72px;"
            >
              <div
                class="grid-view__head"
              >
                 
                <div
                  class="grid-view__column grid-view__column--no-border-right"
                  style="width: 72px;"
                >
                  <!---->
                </div>
                  
                <!---->
              </div>
               
              <div
                class="grid-view__body"
              >
                <div
                  class="grid-view__body-inner"
                >
                  <div
                    class="grid-view__placeholder"
                    style="width: 72px;"
                  >
                    <div
                      class="grid-view__placeholder-groups"
                      style="width: 0px;"
                    />
                     
                  </div>
                   
                  <!---->
                   
                  <div
                    class="grid-view__rows"
                    style="transform: translateY(NaNpx) translateX(0px); left: 0px;"
                  />
                   
                  <div
                    class="grid-view__row-placeholder"
                  />
                </div>
              </div>
               
              <div
                class="grid-view__foot"
              >
                <div
                  class="grid-view__foot-info"
                >
                  
        gridView.rowCount - 1
      
                </div>
                 
              </div>
            </div>
             
            <div
              style="display: none;"
            >
              <div
                class="grid-view__field-dragging"
                style="width: 0px; left: 0px;"
              />
               
              <div
                class="grid-view__field-target"
                style="left: 0px;"
              />
            </div>
          </div>
           
           
          <div
            class="grid-view__divider"
            style="left: 72px;"
          />
           
          <!---->
           
          <div
            class="grid-view__right"
            style="left: 72px;"
          >
              
            <div
              class="grid-view__inner"
              style="min-width: 1000px;"
            >
              <div
                class="grid-view__head"
              >
                 
                <!---->
                 
                <div
                  class="grid-view__column"
                  filters=""
                  style="width: 200px;"
                >
                  <div
                    class="grid-view__description"
                  >
                    <div
                      class="grid-view__description-icon-container"
                    >
                      <i
                        class="grid-view__description-icon iconoir-text"
                      />
                       
                      <!---->
                    </div>
                     
                    <div
                      class="grid-view__description-name"
                      title="Last name"
                    >
                      <span>
                        
        Last name
      
                      </span>
                    </div>
                     
                    <!---->
                     
                    <span
                      class="grid-view__description-options"
                    >
                      <!---->
                       
                      <!---->
                    </span>
                     
                    <!---->
                     
                    <div
                      class="grid-view__description-width"
                    />
                  </div>
                </div>
                <div
                  class="grid-view__column"
                  filters=""
                  style="width: 200px;"
                >
                  <div
                    class="grid-view__description"
                  >
                    <div
                      class="grid-view__description-icon-container"
                    >
                      <i
                        class="grid-view__description-icon iconoir-align-left"
                      />
                       
                      <!---->
                    </div>
                     
                    <div
                      class="grid-view__description-name"
                      title="Notes"
                    >
                      <span>
                        
        Notes
      
                      </span>
                    </div>
                     
                    <!---->
                     
                    <span
                      class="grid-view__description-options"
                    >
                      <!---->
                       
                      <!---->
                    </span>
                     
                    <!---->
                     
                    <div
                      class="grid-view__description-width"
                    />
                  </div>
                </div>
                <div
                  class="grid-view__column"
                  filters=""
                  style="width: 200px;"
                >
                  <div
                    class="grid-view__description"
                  >
                    <div
                      class="grid-view__description-icon-container"
                    >
                      <i
                        class="grid-view__description-icon baserow-icon-circle-checked"
                      />
                       
                      <!---->
                    </div>
                     
                    <div
                      class="grid-view__description-name"
                      title="Active"
                    >
                      <span>
                        
        Active
      
                      </span>
                    </div>
                     
                    <!---->
                     
                    <span
                      class="grid-view__description-options"
                    >
                      <!---->
                       
                      <!---->
                    </span>
                     
                    <!---->
                     
                    <div
                      class="grid-view__description-width"
                    />
                  </div>
                </div>
                <div
                  class="grid-view__column"
                  filters=""
                  style="width: 200px;"
                >
                  <div
                    class="grid-view__description"
                  >
                    <div
                      class="grid-view__description-icon-container"
                    >
                      <i
                        class="grid-view__description-icon iconoir-text"
                      />
                       
                      <!---->
                    </div>
                     
                    <div
                      class="grid-view__description-name"
                      title="Name"
                    >
                      <span>
                        
        Name
      
                      </span>
                    </div>
                     
                    <!---->
                     
                    <span
                      class="grid-view__description-options"
                    >
                      <!---->
                       
                      <!---->
                    </span>
                     
                    <!---->
                     
                    <div
                      class="grid-view__description-width"
                    />
                  </div>
                </div>
                 
                <!---->
              </div>
               
              <div
                class="grid-view__body"
              >
                <div
                  class="grid-view__body-inner"
                >
                  <div
                    class="grid-view__placeholder"
                    style="width: 800px;"
                  >
                    <!---->
                     
                    <div
                      class="grid-view__placeholder-column"
                      style="left: 799px;"
                    />
                    <div
                      class="grid-view__placeholder-column"
                      style="left: 199px;"
                    />
                    <div
                      class="grid-view__placeholder-column"
                      style="left: 399px;"
                    />
                    <div
                      class="grid-view__placeholder-column"
                      style="left: 599px;"
                    />
                  </div>
                   
                  <!---->
                   
                  <div
                    class="grid-view__rows"
                    style="transform: translateY(NaNpx) translateX(0px); left: 0px;"
                  />
                   
                  <div
                    class="grid-view__row-placeholder"
                  />
                </div>
              </div>
               
              <div
                class="grid-view__foot"
              >
                <!---->
                 
                <div
                  style="width: 200px;"
                >
                  <div
                    class="grid-view-aggregation"
                  >
                    <div
                      class="grid-view-aggregation__empty"
                    >
                      <i
                        class="grid-view-aggregation__empty-icon iconoir-plus"
                      />
                      
    common.summarize
  
                    </div>
                     
                  </div>
                </div>
                <div
                  style="width: 200px;"
                >
                  <div
                    class="grid-view-aggregation"
                  >
                    <div
                      class="grid-view-aggregation__empty"
                    >
                      <i
                        class="grid-view-aggregation__empty-icon iconoir-plus"
                      />
                      
    common.summarize
  
                    </div>
                     
                  </div>
                </div>
                <div
                  style="width: 200px;"
                >
                  <div
                    class="grid-view-aggregation"
                  >
                    <div
                      class="grid-view-aggregation__empty"
                    >
                      <i
                        class="grid-view-aggregation__empty-icon iconoir-plus"
                      />
                      
    common.summarize
  
                    </div>
                     
                  </div>
                </div>
                <div
                  style="width: 200px;"
                >
                  <div
                    class="grid-view-aggregation"
                  >
                    <div
                      class="grid-view-aggregation__empty"
                    >
                      <i
                        class="grid-view-aggregation__empty-icon iconoir-plus"
                      />
                      
    common.summarize
  
                    </div>
                     
                  </div>
                </div>
              </div>
            </div>
             
            <div
              style="display: none;"
            >
              <div
                class="grid-view__field-dragging"
                style="width: 0px; left: 0px;"
              />
               
              <div
                class="grid-view__field-target"
                style="left: 0px;"
              />
            </div>
          </div>
           
          <div
            class="grid-view__row-dragging-container"
            style="left: 0px; display: none;"
          >
            <div
              class="grid-view__row-dragging"
              style="width: 872px; top: 0px;"
            />
             
            <div
              class="grid-view__row-target"
              style="width: 872px; top: 0px;"
            />
          </div>
           
           
        </div>
         
        <!---->
      </div>
    </div>
  </div>
</div>
`;
